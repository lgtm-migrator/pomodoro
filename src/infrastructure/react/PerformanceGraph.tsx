import React from "react";
import { ResponsiveContainer, XAxis, YAxis, Bar, BarChart } from "recharts";
import styled from "styled-components";
import { DateSerializer } from "../../domain/date-serializer";
import { IPerformanceGraph } from "../../application/performance-graph";

const Container = styled.div`
  width: 80vw;
  height: 80vh;
`;

export interface IProps {
  performanceGraph: IPerformanceGraph;
}

export const PerformanceGraph = ({ performanceGraph }: IProps) => (
  <Container>
    <ResponsiveContainer>
      <BarChart data={performanceGraph.points}>
        <XAxis
          dataKey="date"
          tickFormatter={date =>
            DateSerializer.deserialize(date).toLocaleDateString()
          }
          tickMargin={10}
        />
        <YAxis tickMargin={5} />
        <Bar dataKey="pomodoros" fill="salmon" />
      </BarChart>
    </ResponsiveContainer>
  </Container>
);
