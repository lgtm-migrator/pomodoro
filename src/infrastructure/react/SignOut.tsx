import { MdExitToApp } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  signOut: () => void;
}

export const SignOut = ({ signOut }: IProps) => (
  <CircleButton onClick={signOut} secondary={true}>
    <MdExitToApp />
  </CircleButton>
);
