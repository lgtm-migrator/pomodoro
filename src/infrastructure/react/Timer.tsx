import React from "react";
import styled from "styled-components";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { TextButton } from "./TextButton";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const ButtonsContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  > * {
    margin: 1em;
  }
`;

const Time = styled.div`
  color: white;
`;

const Minutes = styled.span`
  font-size: 10em;
`;

const Seconds = styled.span`
  font-size: 6em;
  margin-left: 0.2ex;
`;

export interface IProps {
  seconds: number;
  stopped: boolean;
  startTimer: () => void;
  stopTimer: () => void;
  state: PomodoroTimerState;
}

export const Timer = ({
  stopped,
  stopTimer,
  seconds,
  startTimer,
  state
}: IProps) => (
  <Container>
    {stopped ? (
      <ButtonsContainer>
        <TextButton
          onClick={startTimer}
          secondary={state !== PomodoroTimerState.Pomodoro}
        >
          Start{" "}
          {state === PomodoroTimerState.Pomodoro
            ? "Pomodoro"
            : state === PomodoroTimerState.ShortBreak
            ? "Short Break"
            : "Long Break"}
        </TextButton>
      </ButtonsContainer>
    ) : (
      <>
        <Time>
          <Minutes>{Math.floor(seconds / 60)}</Minutes>
          <Seconds>{seconds % 60}</Seconds>
        </Time>
        <TextButton onClick={stopTimer} secondary={true}>
          Stop
        </TextButton>
      </>
    )}
  </Container>
);
