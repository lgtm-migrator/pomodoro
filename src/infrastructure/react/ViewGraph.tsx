import { IoMdStats } from "react-icons/io";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  viewGraph: () => void;
}

export const ViewGraph = ({ viewGraph }: IProps) => (
  <CircleButton secondary={true} onClick={viewGraph}>
    <IoMdStats />
  </CircleButton>
);
