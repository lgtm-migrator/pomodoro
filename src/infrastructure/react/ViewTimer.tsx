import { MdTimer } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  viewTimer: () => void;
}

export const ViewTimer = ({ viewTimer }: IProps) => (
  <CircleButton secondary={true} onClick={viewTimer}>
    <MdTimer />
  </CircleButton>
);
