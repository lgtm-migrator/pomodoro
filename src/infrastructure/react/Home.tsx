import React from "react";
import styled from "styled-components";
import { SignOut } from "./SignOut";
import { Timer } from "./Timer";

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps {
  signOut: () => Promise<void>;
}

export const Home = ({ signOut }: IProps) => {
  return (
    <Container>
      <Timer />
      <SignOutContainer>
        <SignOut signOut={signOut} />
      </SignOutContainer>
    </Container>
  );
};
