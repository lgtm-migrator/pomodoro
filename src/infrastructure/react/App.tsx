import { useAsync } from "react-use";
import { PulseLoader } from "react-spinners";
import { observer } from "mobx-react";
import React from "react";
import styled from "styled-components";
import { AuthenticationStore } from "../mobx/authentication-store";
import { PerformanceGraphStore } from "../mobx/performance-graph-store";
import { PomodoroTimerStore } from "../mobx/pomodoro-timer-store";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps
  extends Pick<
      IHomeProps,
      "stopTimer" | "startTimer" | "signOut" | "viewGraph"
    >,
    ILandingProps {
  authenticationStore: AuthenticationStore;
  pomodoroTimerStore: PomodoroTimerStore;
  performanceGraphStore: PerformanceGraphStore;
  initialize: () => Promise<void>;
}

export const App = observer(
  ({
    authenticationStore: { signedIn },
    pomodoroTimerStore: { seconds, state, stopped },
    performanceGraphStore: { performanceGraph },
    initialize,
    repositoryURL,
    signIn,
    ...homeProps
  }: IProps) => {
    useAsync(initialize, []);

    return signedIn === null ? (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    ) : signedIn ? (
      <Home
        {...homeProps}
        performanceGraph={performanceGraph}
        seconds={seconds}
        state={state}
        stopped={stopped}
      />
    ) : (
      <Landing repositoryURL={repositoryURL} signIn={signIn} />
    );
  }
);
