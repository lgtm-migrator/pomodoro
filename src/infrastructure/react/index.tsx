import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { PerformanceGraphViewer } from "../../application/performance-graph-viewer";
import { PomodoroTimerStopper } from "../../application/pomodoro-timer-stopper";
import { PomodoroTimerStarter } from "../../application/pomodoro-timer-starter";
import { PomodoroTimerStore } from "../mobx/pomodoro-timer-store";
import { AuthenticationStore } from "../mobx/authentication-store";
import { PerformanceGraphStore } from "../mobx/performance-graph-store";
import { App } from "./App";
import { GlobalStyle } from "./style";

export class ReactRenderer {
  constructor(
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly performanceGraphViewer: PerformanceGraphViewer,
    private readonly pomodoroTimerStarter: PomodoroTimerStarter,
    private readonly pomodoroTimerStopper: PomodoroTimerStopper,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly authenticationStore: AuthenticationStore,
    private readonly pomodoroTimerStore: PomodoroTimerStore,
    private readonly performanceGraphStore: PerformanceGraphStore,
    private readonly repositoryURL: string
  ) {}

  public render(element: HTMLElement): void {
    render(
      <>
        <App
          authenticationStore={this.authenticationStore}
          pomodoroTimerStore={this.pomodoroTimerStore}
          performanceGraphStore={this.performanceGraphStore}
          initialize={() => this.applicationInitializer.initialize()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          stopTimer={() => this.pomodoroTimerStopper.stop()}
          startTimer={() => this.pomodoroTimerStarter.start()}
          viewGraph={() => this.performanceGraphViewer.viewGraph()}
        />
        <GlobalStyle />
      </>,
      element
    );
  }
}
