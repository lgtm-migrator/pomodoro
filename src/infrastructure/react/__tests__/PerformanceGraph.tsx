import { render } from "@testing-library/react";
import React from "react";
import { PerformanceGraph } from "../PerformanceGraph";
import { DateSerializer } from "../../../domain/date-serializer";

it("renders with no data", () => {
  const result = render(
    <div>
      <PerformanceGraph performanceGraph={{ points: [] }} />
    </div>
  );
  expect(result.container).toMatchSnapshot();
});

it("renders with data", () => {
  const result = render(
    <div>
      <PerformanceGraph
        performanceGraph={{
          points: [
            { date: DateSerializer.serialize(new Date(0)), pomodoros: 42 }
          ]
        }}
      />
    </div>
  );
  expect(result.container).toMatchSnapshot();
});
