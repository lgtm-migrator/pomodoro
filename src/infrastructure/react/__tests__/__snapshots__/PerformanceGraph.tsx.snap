// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with data 1`] = `
<div>
  <div>
    <div
      class="sc-bdVaJa idVYWg"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 100%;"
      >
        <div
          style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with no data 1`] = `
<div>
  <div>
    <div
      class="sc-bdVaJa idVYWg"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 100%;"
      >
        <div
          style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
        />
      </div>
    </div>
  </div>
</div>
`;
